File: elixir/pythagorean-triplet/pythagorean_triplet.exs
defmodule Triplet do

  @doc """
  Calculates sum of a given triplet of integers.
  """
  @spec sum([non_neg_integer]) :: non_neg_integer
  def sum(triplet) do
    Enum.sum(triplet)
  end

  @doc """
  Calculates product of a given triplet of integers.
  """
  @spec product([non_neg_integer]) :: non_neg_integer
  def product(triplet) do
    Enum.reduce(triplet, 1, &(&1 * &2))
  end

  @doc """
  Determines if a given triplet is pythagorean. That is, do the squares of a and b add up to the square of c?
  """
  @spec pythagorean?([non_neg_integer]) :: boolean
  def pythagorean?([a, b, c]) do
    (a*a) + (b*b) == (c*c)
  end

  @doc """
  Generates a list of pythagorean triplets from a given min (or 1 if no min) to a given max.
  """
  @spec generate(non_neg_integer, non_neg_integer) :: [list(non_neg_integer)]
  def generate(min \\ 1, max) do
    # This is pretty brute-force, and it comes up with dupes, but it works.
    all = for a <- Range.new(min, max-2),
              b <- Range.new(min+1, max-1),
              c <- Range.new(min+2, max),
              pythagorean?([a, b, c]),
          into: [],
          do: [a, b, c]
    all
    |> Enum.map(&Enum.sort/1)
    |> Enum.uniq
  end

  @doc """
  Generates a list of pythagorean triplets from a given min to a given max, whose values add up to a given sum.
  """
  @spec generate(non_neg_integer, non_neg_integer, non_neg_integer) :: [list(non_neg_integer)]
  def generate(min, max, sum) do
    generate(min, max)
    |> Enum.filter(&(sum(&1) == sum))
  end
end

File: javascript/pythagorean-triplet/pythagorean-triplet.js
var Triplet = function(a, b, c) {
  this.trips = [a, b, c];
};

Triplet.prototype.sum = function() {
  return this.trips.reduce((p, c) => p + c, 0);
};

Triplet.prototype.product = function() {
  return this.trips.reduce((p, c) => p * c, 1);
};

Triplet.prototype.isPythagorean = function() {
  let [a, b, c] = this.trips;
  return (a*a) + (b*b) == (c*c);
};

Triplet.where = function(params) {
  let min = params.minFactor || 1;
  let max = params.maxFactor;
  let sum = params.sum || false;
  let out = [];
  for (var a=min; a<max-1; a++) {
    for (var b=a+1; b<max; b++) {
      for (var c=b+1; c<=max; c++) {
        let t = new Triplet(a,b,c);
        if (t.isPythagorean()) {
          if (!sum) {
            out.push(t);
          }
          if (sum && sum == t.sum()) {
            out.push(t);
          }
        }
      }
    }
  }
  return out;
};

module.exports = Triplet;
